[MANA-133] Fix MPI_REQUEST_NULL not being skipped in Testany

fyshhh opened the pull request:
This pull requests fixes a bug where `MPI_REQUEST_NULL` handles were not being skipped properly in `MPI_Testany`.

gc00 commented on mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:
```diff
@@ -156,9 +156,14 @@ USER_DEFINED_WRAPPER(int, Testany, (int) count,
   MPI_Status *local_status = status;
 
   int retval;
+  bool all_requests_null = true;
```
I think there is a simpler logic that doesn't need the variable `all_requests_null`.  I propose removing this variable, and making the changes noted below.

But if we must change the name, then this is better:
  `all_requests_null` -> `all_requests_are_null`

gc00 commented on mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:
```diff
@@ -156,9 +156,14 @@ USER_DEFINED_WRAPPER(int, Testany, (int) count,
 
   int retval;
+  bool all_requests_null = true;
   *local_flag = 0;
```
Instead, let's initialize it as:  `*local_flag = 1;`

gc00 commented on mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:
```diff
@@ -156,9 +156,14 @@ USER_DEFINED_WRAPPER(int, Testany, (int) count,
+    if (local_array_of_requests[i] == MPI_REQUEST_NULL) {
+      continue;
+    }
+    all_requests_null = false;
```
Delete the line above.  We don't need it now.

gc00 commented on mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:
```diff
@@ -168,6 +173,9 @@ USER_DEFINED_WRAPPER(int, Testany, (int) count,
       break;
     }
   }
```
Delete the next three lines:  `if (all_requests_null) { ... }`.

If `MPI_Test` was previously called and failed, then `MPI_Test` will have already set `*local_flag = 0;`.
If `MPI_Test` was never called, then `*local_flag` still has its initial value of `1`, which is the correct value when `all_requests_are_null` is true.

gc00 reviewed (commented):
@illio, Nice catch.

However, in my comments, I'm suggesting a simpler logic for doing the same thing.  We don't need `all_requests_null`.  Instead, we can simply initialize `*local_flag` to `1`.  This does the right thing when `all_requests_null` is true.

Could you make these changes, and then I'll look at it again?  Thanks.

fyshhh commented:
@gc00 thanks for your suggestions. I've removed `all_requests_null` and initialised `*local_flag` to 1. The only difference is that I added `*local_flag = 0` before each `MPI_Test` call. That seemed to be the necessary piece to tie it together - it seems like `*local_flag` was not being set to 0 in some cases.

gc00 commented on mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:
```diff
@@ -167,9 +167,13 @@ USER_DEFINED_WRAPPER(int, Testany, (int) count,
+    if (local_array_of_requests[i] == MPI_REQUEST_NULL) {
+      continue;
+    }
+    *local_flag = 0;
```
In principle, we should not need the line above (setting `*local_flag = 0;`).  I'd feel better if you can remove this line, and run it again to see why it's needed.  Logically, this should not change the result, since `MPI_Test` must also set `*local_flag`.  And it's dangerous to commit code that we don't understand.

fyshhh replied:
I re-ran the test and it works without this line - I probably made a mistake in the prior tests.

gc00 reviewed (approved):
Please see why `*local_flag = 0` is needed, before pushing this in.  If it is needed, I'm worried that it indicates memory corruption or some other underlying issue.
